### PR TITLE
Move unit tests in a dedicated subdir

### PR DESCRIPTION
**Pull request by fbuihuu**

### CODE REVIEW

**Comment by fbuihuu** [comment]
@bluca it is a follow-up for https://github.com/systemd/systemd/pull/26221.

https://salsa.debian.org/systemd-team/systemd/-/blob/upstream-ci/debian/tests/unit-tests needs be be adjusted again because `run-unit-tests.py` has been moved to `/usr/lib/systemd/tests/unit-tests` again.

Maybe we can adjust the script to find `run-unit-tests.py` under `/usr/lib/systemd/tests` instead of hard coding its path. So it's less subject to break when `run-unit-tests.py` is moved.

WDYT ?

**Comment by bluca** [comment]
Hopefully we won't move it that often :-) I could simply look in either location?

**Comment by fbuihuu** [comment]
Who knows... ;) But yeah looking in either location would work too.

File: meson_options.txt [hunk 1]
```diff
@@ -472,6 +472,8 @@ option('fuzz-tests', type : 'boolean', value : false,
        description : 'run the fuzzer regression tests by default (with sanitizers)')
 option('install-tests', type : 'boolean', value : false,
        description : 'install test executables')
+option('unittestsdir', type : 'string', value : '',
```
Comment (bluca) [inline meson_options.txt]: we shouldn't need an option for this, right?

**Comment by bluca** [comment]
updated

File: meson_options.txt (hunk 1 above)
Comment (fbuihuu) [reply meson_options.txt]: oops yeah a leftover... good catch.

**Comment by fbuihuu** [comment]
@bluca @keszybz there are 2 packaging issues (at least) with this change.

Debian:
```
autopkgtest [17:55:55]: test udev: [----------------------- 24:28
cp: cannot stat '/usr/lib/systemd/tests/manual/test-udev': No such file or directory
```

Fedora:
```
+ sed -i 's|#!/usr/bin/env python3|#!/usr/bin/python3|' /builddir/build/BUILDROOT/systemd-253-1.20230320173402231036.pr26870.810.g4e52fe08fa.x86_64/usr/lib/systemd/tests/run-unit-tests.py
sed: can't read /builddir/build/BUILDROOT/systemd-253-1.20230320173402231036.pr26870.810.g4e52fe08fa.x86_64/usr/lib/systemd/tests/run-unit-tests.py: No such file or directory
```

If it makes sense to accept this PR, how these issues should be fixed ? before accepting the PR ?

File: test/meson.build [hunk 2]
```diff
@@ -85,6 +85,12 @@ if install_tests
         install_data('create-busybox-container',
                      install_mode : 'rwxr-xr-x',
                      install_dir : testdata_dir)
+
+        # The unit tests implemented as shell scripts expect to find testdata/
+        # in the directory where they are stored.
+        meson.add_install_script(meson_make_symlink,
+                                testdata_dir,
+                                unittestsdir / 'testdata')
```
Comment (keszybz) [inline test/meson.build]: Indentation.

**Comment by keszybz** [comment]
So… what about **not** moving `run-unit-tests.py`? This way we don't need to adjust packaging, and it actually nicer, because `run-unit-tests.py` already has "unit tests" in the name and we avoid a duplication with the directory name and the user doesn't need to descend into a subdirectory. (Maybe we can have `/usr/lib/systemd/tests/run-integration-tests.{py,sh}` in the future.)

**Comment by bluca** [comment]
> @bluca @keszybz there are 2 packaging issues (at least) with this change.
> 
> Debian:
> 
> ```
> autopkgtest [17:55:55]: test udev: [----------------------- 24:28
> cp: cannot stat '/usr/lib/systemd/tests/manual/test-udev': No such file or directory
> ```

I already fixed this yesterday

**Comment by keszybz** [comment]
@fbuihuu did you see my comment right before your push?

**Comment by fbuihuu** [comment]
@keszybz sorry I couldn't reply until now.

I don't have any objection to not move `run-unit-tests.py`. I'll update this PR.

**Comment by keszybz** [comment]
LGTM. With this tweak, the tests should pass.

**Comment by keszybz** [comment]
CentOS CI (CentOS Stream 8) fails in TEST-02-UNITTESTS. I'm not sure if this is related.

**Comment by fbuihuu** [comment]
This one:

```
=== Failed test log ===
--- test-bus-marshal begin ---
/usr/lib/systemd/tests/unit-tests/test-bus-marshal: error while loading shared libraries: libgio-2.0.so.0: cannot open shared object file: No such file or directory
```

looks unrelated I think.

**Comment by mrc0mmand** [comment]
> This one:
> 
> ```
> === Failed test log ===
> --- test-bus-marshal begin ---
> /usr/lib/systemd/tests/unit-tests/test-bus-marshal: error while loading shared libraries: libgio-2.0.so.0: cannot open shared object file: No such file or directory
> ```
> 
> looks unrelated I think.

It fails with the same error on both C8S and Arch, so I'd say it's related.

**Comment by fbuihuu** [comment]
@mrc0mmand it seems that `libgio` is not installed in the target image for CentOS CI whereas `test-bus-marshal` depends on this lib. Given the fact that the lib seems to be installed on the host  I have no idea why it's missing in the image...

**Comment by mrc0mmand** [comment]
> @mrc0mmand it seems that `libgio` is not installed in the target image for CentOS CI whereas `test-bus-marshal` depends on this lib. Given the fact that the lib seems to be installed on the host I have no idea why it's missing in the image...

I guess this beauty needs to be updated as well:

https://github.com/systemd/systemd/blob/c5ba7a2a4dd19a2d31b8a9d52d3c4bdde78387f0/test/test-functions#L1272-L1278

**Comment by keszybz** [comment]
> beauty

Ack.

**Comment by fbuihuu** [comment]
@mrc0mmand thanks a lot for the hint. I'll update the PR.

**Review by mrc0mmand — APPROVED** [review]

**Comment by bluca** [comment]
so should I drop the change to look for the run script in the subdir?

**Comment by fbuihuu** [comment]
Yes please.

**Comment by keszybz** [comment]
@bluca do you want to restart the jobs? We could merge this now.

**Comment by bluca** [comment]
> @bluca do you want to restart the jobs? We could merge this now.

they are still pending, did you restart them? let's make sure it works before merging

**Comment by bluca** [comment]
This failure in TEST-02-UNITTESTS looks related:

```
++ mktemp --tmpdir --directory test-network-generator-conversion.XXXXXXXXXX
+ out=/tmp/test-network-generator-conversion.cQHjKwWZbL
+ trap 'rm -rf '\''/tmp/test-network-generator-conversion.cQHjKwWZbL'\''' EXIT INT QUIT PIPE
++ cat '/usr/lib/systemd/tests/unit-tests/testdata/test-network-generator-conversion/test-*.input'
cat: '/usr/lib/systemd/tests/unit-tests/testdata/test-network-generator-conversion/test-*.input': No such file or dire
ctory
```

**Comment by fbuihuu** [comment]
It seems there's something wrong with `/usr/lib/systemd/tests/unit-tests/testdata`. It's supposed to be a symlink created by:
```
Running custom install script '/tmp/autopkgtest.SYMA33/build.lDR/systemd/tools/meson-make-symlink.sh /usr/lib/systemd/tests/testdata /usr/lib/systemd/tests/unit-tests/testdata'
```
Is there any way to check if it's correctly installed and it's content is `../testdata` on this platform ?

**Comment by bluca** [comment]
You'll need to build the image on a jammy host/vm to check, or add some debug prints to the test

**Comment by fbuihuu** [comment]
```
+ ls -l /usr/lib/systemd/tests/unit-tests/testdata
ls: cannot access '/usr/lib/systemd/tests/unit-tests/testdata': No such file or directory
```
@bluca it seems that there's something in the packaging that dropped the symlink. Any idea what it could be ?

**Comment by fbuihuu** [comment]
The symlink is supposed to be created by the following commit in this PR: https://github.com/systemd/systemd/pull/26870/commits/300c42e9c908bf61aeb44c7d08c6c18fe6014a8b

**Comment by bluca** [comment]
The symlink is in the package - for reference, at the same URL where the test run log is stored, there's also always an `artifacts.tar.gz` archive that has journal files and the built packages that are used for the tests. The symlink and directory are there:

```
$ ls -l usr/lib/systemd/tests/unit-tests/testdata
lrwxrwxrwx 1 bluca bluca 11 Mar 23 19:07 usr/lib/systemd/tests/unit-tests/testdata -> ../testdata
$ ls -l usr/lib/systemd/tests/unit-tests/testdata/
total 4
-rwxr-xr-x  1 bluca bluca 1394 Jun  3  2022 create-busybox-container
drwxr-xr-x  2 bluca bluca   80 Jun  3  2022 journal-data
drwxr-xr-x  2 bluca bluca  240 Jun  3  2022 test-bcd
drwxr-xr-x  2 bluca bluca 3180 Jun  3  2022 test-execute
drwxr-xr-x 19 bluca bluca  720 Jun  3  2022 test-fstab-generator
drwxr-xr-x  2 bluca bluca   60 Jun  3  2022 test-keymap-util
drwxr-xr-x  5 bluca bluca  160 Jun  3  2022 test-network-generator-conversion
drwxr-xr-x  2 bluca bluca  380 Jun  3  2022 test-path
drwxr-xr-x  2 bluca bluca   60 Jun  3  2022 test-path-util
drwxr-xr-x  2 bluca bluca  400 Jun  3  2022 test-resolve
drwxr-xr-x  2 bluca bluca  160 Jun  3  2022 testsuite-03.units
drwxr-xr-x  2 bluca bluca  120 Jun  3  2022 testsuite-04.units
drwxr-xr-x  2 bluca bluca   80 Jun  3  2022 testsuite-06.units
drwxr-xr-x  3 bluca bluca  120 Jun  3  2022 testsuite-08.units
drwxr-xr-x  2 bluca bluca   80 Jun  3  2022 testsuite-10.units
drwxr-xr-x  2 bluca bluca   60 Jun  3  2022 testsuite-11.units
drwxr-xr-x  2 bluca bluca  200 Jun  3  2022 testsuite-16.units
drwxr-xr-x  2 bluca bluca  100 Jun  3  2022 testsuite-28.units
drwxr-xr-x  3 bluca bluca   60 Jun  3  2022 testsuite-30.units
drwxr-xr-x  2 bluca bluca   80 Jun  3  2022 testsuite-52.units
drwxr-xr-x  2 bluca bluca  120 Jun  3  2022 testsuite-63.units
drwxr-xr-x  2 bluca bluca 1280 Jun  3  2022 test-sysusers
drwxr-xr-x  2 bluca bluca  120 Jun  3  2022 test-umount
drwxr-xr-x  7 bluca bluca 5480 Jun  3  2022 units
```

**Comment by bluca** [comment]
Are you sure the install scripts actually installs the symlink?

**Comment by fbuihuu** [comment]
> Are you sure the install scripts actually installs the symlink?

I'm not sure to follow since in https://github.com/systemd/systemd/pull/26870#issuecomment-1481752481, you actually said that the symlink is part of the package.

So it's was in the package but it's been lost after the package was installed ?

The symlink is perfectly installed on my testing with openSUSE TW.

**Comment by bluca** [comment]
The way this works is that the host is provisioned installing the packages, and then the guest image is assembled using the local scripts in test/. Downloading the package shows that everything is in order, and that's why I was asking if there could be issues in the following step

**Comment by fbuihuu** [comment]
Ok I see now what you meant.

Let's see if I can spot something in this function by adding some extra debug logs:

```
install_debian_systemd() {
    dinfo "Install debian systemd"
    
    local files
    
    while read -r deb; do
        files="$(dpkg-query -L "$deb" 2>/dev/null)" || continue
        ddebug "Install debian files from package $deb"
        for file in $files; do
            [ -e "$file" ] || continue 
            [ -d "$file" ] && continue
            inst "$file"
        done
    done < <(grep -E '^Package:' "${SOURCE_DIR}/debian/control" | cut -d ':' -f 2)
}   

```
the logs show that it's not considering the symlink.

**Comment by bluca** [comment]
```
D: Considering /usr/lib/systemd/tests/unit-tests/testdata
drwxr-xr-x 27 root root 4096 Mar 24 16:44 /usr/lib/systemd/tests/testdata
lrwxrwxrwx 1 root root 11 Jun  3  2022 /usr/lib/systemd/tests/unit-tests/testdata -> ../testdata
/usr/lib/systemd/tests/testdata
```

**Review by bluca — APPROVED** [review]

**Comment by fbuihuu** [comment]
>         for file in $files; do
>             [ -e "$file" ] || continue 
>             [ -d "$file" ] && continue

So the (hopefully) last issue was that the conditional expression follow symlinks.

@bluca thanks for helping me solving this issue.

**Pull request merged**